ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenRunner.java

```diff
@@ -184,45 +190,107 @@ public Void visitLikePredicate(final LikePredicate node, final TypeContext conte
     @Override
     public Void visitFunctionCall(final FunctionCall node, final TypeContext context) {
       final FunctionName functionName = node.getName();
-
-      // this context gets updated as we process non lambda arguments
-      final TypeContext currentTypeContext = context.getCopy();
-
-      final List<SqlArgument> argumentTypes = new ArrayList<>();
+      final List<SqlArgument> argumentSchemas = new ArrayList<>();
       final List<TypeContext> typeContextsForChildren = new ArrayList<>();
-      final boolean hasLambda = node.hasLambdaFunctionCallArguments();
-      for (final Expression argExpr : node.getArguments()) {
-        final TypeContext childContext = TypeContextUtil.contextForExpression(
-            argExpr, context, currentTypeContext
-        );
-        typeContextsForChildren.add(childContext);
-
-        // pass a copy of the context to the type checker so that type checking in one
-        // expression subtree doesn't interfere with type checking in another one
-        final SqlType resolvedArgType =
-            expressionTypeManager.getExpressionSqlType(argExpr, childContext.getCopy());
-
-        if (argExpr instanceof LambdaFunctionCall) {
-          argumentTypes.add(
+      final List<SqlArgument> argTypesForFirstPass = new ArrayList<>();
+      final List<Expression> arguments = node.getArguments();
+
+      for (final Expression expression : arguments) {
```

Comment:
nit: it would be better to add a high-level comment describing the two-pass algorithm here, e.g. why and how we do it.

Reply:
added that as a java doc as part of the new util function used in all 3 classes